### ISSUE TITLE
Replace bower-asset with npm-asset

### ISSUE THREAD

**Issue by karelvasicek** [yiisoft/yii2#20067]
Related f.ex. to https://github.com/hiqdev/asset-packagist/issues/163 and especially to https://github.com/hiqdev/asset-packagist/issues/163#issuecomment-1805642726 .

`composer update` results to some packages downgrade.

Question is if/how it's related to https://bower.io/, it (as stated there) is maintained, but f.ex.
  - it's officially stated that "...we recommend using Yarn and Vite..."
  - package search on https://bower.io/search/ does not work at all

All packages required by Yii2 are available as `npm-asset/package` and versions seem to be correct in that case.

**Comment by schmunk42** [comment]
Just replacing bower with npm might lead to problems, since both services work very differently.
In a nutshell: Bower was actually made in a way how we use it in Yii 2 for (pre-bundled) frontend assets, npm is more meant to be build with ie. webpack, although most packages have dist files, but it's not required. Also the dependencies may vary a lot.

I'd strongly recommend to wait for a fix in asset-packagist.

@bizley @terabytesoftw A change from bower to npm might be an option for 2.2, but please not for 2.0.

**Comment by r-simlinger** [comment]
Is there any workaround?
Does not look like there will be someone wo will fix it from the asset-packagist side.

**Comment by ItsReddi** [comment]
A the moment bower-asset/jquery is not installable anymore for us. @karelvasicek do you have any workaround?

**Comment by karelvasicek** [comment]
@ItsReddi 

> do you have any workaround?

No, we unfortunately don't have any. We updated our main project some time ago and it hadn't worked either initially (packages had been downgraded), but started to work some time later (packages were updated again). It seems to work from time to time.

**Comment by ailmanki** [comment]
In composer.json:
add (this makes composer believe these packages are installed)
``` "replace": {
    "bower-asset/jquery": "*",
    "bower-asset/jquery-ui": "*"
}
```

Next install the depencies with npm,

```
 "jquery-ui": "npm:components-jqueryui@^1.12.1",
 "jquery": "^3.0.0"
```
 
Above dependency for jquery-ui, makes it being installed into jquery-ui instead of components-jqueryui, long ago - I think that fixes some issues with Yii and maybe older JS.
 
 Lastly see here how to tell yii Asset Manager to load your npm files..
 https://www.yiiframework.com/doc/guide/2.0/en/structure-assets#customizing-asset-bundles
 
 
 Rinse and repeat for whatever library u need to replace.

**Comment by ItsReddi** [comment]
@karelvasicek @ailmanki Thanks.
For reference:
https://github.com/hiqdev/asset-packagist/issues/167

**Comment by terabytesoftw** [comment]
Here i think it is worth considering two solutions for the future.

1. One possible solution is to use cdn, both in dev and prod versions.
2. The other is to extract the assets from the composer.json of the packages and use npm, all the packages used are compatible and have dist for use, in this case we can use npm or php-forge/foxy for this purpose.

**Comment by ailmanki** [comment]
It seems to me no issue; the framework allows perfectly to replace it with NPM or any other mechanism.
Maybe a guide would help - instead of my small post above?

**Comment by inversed** [comment]
I don't know what's been going on with Bower's jQuery but I clicked Fetch Updates just now on https://asset-packagist.org/package/bower-asset/jquery and it is now showing the newer releases of jQuery beyond 3.6.1. My composer project is no longer downgrading my bower jquery package.

**Comment by ItsReddi** [comment]
> I don't know what's been going on with Bower's jQuery but I clicked Fetch Updates just now on https://asset-packagist.org/package/bower-asset/jquery and it is now showing the newer releases of jQuery beyond 3.6.1. My composer project is no longer downgrading my bower jquery package.

Sadly that works only sometimes for some hours. It is already gone again.
![grafik](https://github.com/user-attachments/assets/e00028b5-0184-488b-84dd-9ac9d66dc749)

**Comment by samdark** [comment]
@SilverFire, @hiqsol would you please take a look at this one?

**Comment by alireza-salehi** [comment]
we already use yarn in our projects and removed bower completely
